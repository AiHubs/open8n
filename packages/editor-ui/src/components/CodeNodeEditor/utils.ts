import type * as esprima from 'esprima-next';
import type { Completion } from '@codemirror/autocomplete';
import type { Node } from 'estree';
import type { RangeNode } from './types';
import type { INodeUi, Schema } from '@/Interface';
import type { INodeExecutionData } from 'n8n-workflow';
import { sanitizeHtml } from '@/utils/htmlUtils';
import { useWorkflowsStore } from '@/stores/workflows.store';
import { useNDVStore } from '@/stores/ndv.store';
import { useDataSchema } from '@/composables/useDataSchema';
import { executionDataToJson } from '@/utils/nodeTypesUtils';

export function walk<T extends RangeNode>(
	node: Node | esprima.Program,
	test: (node: Node) => boolean,
	found: Node[] = [],
) {
	// @ts-ignore
	if (test(node)) found.push(node);

	for (const key in node) {
		if (!(key in node)) continue;

		// @ts-ignore
		const child = node[key];

		if (child === null || typeof child !== 'object') continue;

		if (Array.isArray(child)) {
			child.forEach((node) => walk(node, test, found));
		} else {
			walk(child, test, found);
		}
	}

	return found as T[];
}

export const escape = (str: string) =>
	str
		.replace('$', '\\$')
		.replace('(', '\\(')
		.replace(')', '\\)')
		.replace('[', '\\[')
		.replace(']', '\\]');

export const toVariableOption = (label: string) => ({ label, type: 'variable' });

export const addVarType = (option: Completion) => ({ ...option, type: 'variable' });

export const addInfoRenderer = (option: Completion): Completion => {
	const { info } = option;
	if (typeof info === 'string') {
		option.info = () => {
			const wrapper = document.createElement('span');
			wrapper.innerHTML = sanitizeHtml(info);
			return wrapper;
		};
	}
	return option;
};

export function getParentNodes() {
	const activeNode = useNDVStore().activeNode;
	const { getCurrentWorkflow, getNodeByName } = useWorkflowsStore();
	const workflow = getCurrentWorkflow();

	if (!activeNode || !workflow) return [];

	return workflow
		.getParentNodesByDepth(activeNode?.name)
		.filter(({ name }, i, nodes) => {
			return name !== activeNode.name && nodes.findIndex((node) => node.name === name) === i;
		})
		.map((n) => getNodeByName(n.name))
		.filter((n) => n !== null);
}

export function getSchemas() {
	const parentNodes = getParentNodes();
	const parentNodesNames = parentNodes.map((node) => node?.name);
	const { getSchemaForExecutionData, getInputDataWithPinned } = useDataSchema();
	const parentNodesSchemas: Array<{ nodeName: string; schema: Schema }> = parentNodes
		.map((node) => {
			const inputData: INodeExecutionData[] = getInputDataWithPinned(node);

			return {
				nodeName: node?.name || '',
				schema: getSchemaForExecutionData(executionDataToJson(inputData), true),
			};
		})
		.filter((node) => node.schema?.value.length > 0);

	const inputSchema = parentNodesSchemas.shift();

	return {
		parentNodesNames,
		inputSchema,
		parentNodesSchemas,
	};
}
