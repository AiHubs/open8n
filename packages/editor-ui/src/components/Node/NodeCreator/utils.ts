import { INodeCreateElement, ActionCreateElement, SubcategorizedNodeTypes } from '@/Interface';
import { CORE_NODES_CATEGORY } from '@/constants';
import { v4 as uuidv4 } from 'uuid';
import { INodeTypeDescription } from 'n8n-workflow';

export function transformNodeType(
	node: INodeTypeDescription,
	subcategory?: string,
	type = 'node',
): INodeCreateElement {
	return {
		uuid: uuidv4(),
		key: node.name,
		subcategory: subcategory ?? node.codex?.subcategories?.[CORE_NODES_CATEGORY]?.[0] ?? '*',
		properties: node,
		type,
	};
}

export function subcategorizeItems(items: INodeTypeDescription[]) {
	return items.reduce((acc: SubcategorizedNodeTypes, item) => {
		// Only Core Nodes subcategories are valid, others are uncategorized
		const isCoreNodesCategory = item.codex?.categories?.includes(CORE_NODES_CATEGORY);
		const subcategories = isCoreNodesCategory
			? item?.codex?.subcategories?.[CORE_NODES_CATEGORY] ?? []
			: ['*'];

		subcategories.forEach((subcategory: string) => {
			if (!acc[subcategory]) {
				acc[subcategory] = [];
			}
			acc[subcategory].push(transformNodeType(item, subcategory));
		});

		return acc;
	}, {});
}

export function sortNodeCreateElements(nodes: INodeCreateElement[]) {
	return nodes.sort((a, b) => {
		if (a.type !== 'node' || b.type !== 'node') return -1;
		const displayNameA = a.properties.displayName.toLowerCase();
		const displayNameB = b.properties.displayName.toLowerCase();

		return displayNameA.localeCompare(displayNameB, undefined, { sensitivity: 'base' });
	});
}
