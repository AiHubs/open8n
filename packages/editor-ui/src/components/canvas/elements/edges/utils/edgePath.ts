import type { EdgeProps } from '@vue-flow/core';
import { getBezierPath, getSmoothStepPath, Position } from '@vue-flow/core';
import { NodeConnectionType } from 'n8n-workflow';

const EDGE_PADDING_BOTTOM = 130;
const EDGE_PADDING_X = 40;
const EDGE_BORDER_RADIUS = 16;
const HANDLE_SIZE = 20; // Required to avoid connection line glitching when initially interacting with the handle

const isRightOfSourceHandle = (sourceX: number, targetX: number) => sourceX - HANDLE_SIZE > targetX;

export function getCustomPath(
	props: Pick<
		EdgeProps,
		'sourceX' | 'sourceY' | 'sourcePosition' | 'targetX' | 'targetY' | 'targetPosition'
	>,
	{
		connectionType = NodeConnectionType.Main,
	}: {
		connectionType?: NodeConnectionType;
	} = {},
): string[] {
	const { targetX, targetY, sourceX, sourceY, sourcePosition, targetPosition } = props;

	if (!isRightOfSourceHandle(sourceX, targetX) || connectionType !== NodeConnectionType.Main) {
		const [segment] = getBezierPath(props);
		return [segment];
	}

	// Connection is backwards and the source is on the right side
	// -> We need to avoid overlapping the source node
	const firstSegmentTargetX = sourceX;
	const firstSegmentTargetY = sourceY + EDGE_PADDING_BOTTOM;
	const [firstSegmentPath] = getSmoothStepPath({
		sourceX,
		sourceY,
		targetX: firstSegmentTargetX,
		targetY: firstSegmentTargetY,
		sourcePosition,
		targetPosition: Position.Right,
		borderRadius: EDGE_BORDER_RADIUS,
		offset: EDGE_PADDING_X,
	});

	const [secondSegmentPath] = getSmoothStepPath({
		sourceX: firstSegmentTargetX,
		sourceY: firstSegmentTargetY,
		targetX,
		targetY,
		sourcePosition: Position.Left,
		targetPosition,
		borderRadius: EDGE_BORDER_RADIUS,
		offset: EDGE_PADDING_X,
	});

	return [firstSegmentPath, secondSegmentPath];
}
