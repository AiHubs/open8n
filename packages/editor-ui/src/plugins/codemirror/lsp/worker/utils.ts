import type { Schema } from '@/Interface';
import type { Diagnostic } from '@codemirror/lint';
import { capitalize } from 'lodash-es';
import { type DocMetadata, type DocMetadataArgument } from 'n8n-workflow';
import ts from 'typescript';

export const FILE_NAME = 'index.js';
const FN_PREFIX = '(() => {\n';

export function wrapInFunction(script: string): string {
	return `${FN_PREFIX}${script}\n})()`;
}

export function cmPosToTs(pos: number) {
	return pos + FN_PREFIX.length;
}

export function tsPosToCm(pos: number) {
	return pos - FN_PREFIX.length;
}

/**
 * TypeScript has a set of diagnostic categories,
 * which maps roughly onto CodeMirror's categories.
 * Here, we do the mapping.
 */
export function tsCategoryToSeverity(
	diagnostic: Pick<ts.DiagnosticWithLocation, 'category' | 'code'>,
): Diagnostic['severity'] {
	if (diagnostic.code === 7027) {
		// Unreachable code detected
		return 'warning';
	}
	switch (diagnostic.category) {
		case ts.DiagnosticCategory.Error:
			return 'error';
		case ts.DiagnosticCategory.Message:
			return 'info';
		case ts.DiagnosticCategory.Warning:
			return 'warning';
		case ts.DiagnosticCategory.Suggestion:
			return 'info';
	}
}

/**
 * Not all TypeScript diagnostic relate to specific
 * ranges in source code: here we filter for those that
 * do.
 */
export function isDiagnosticWithLocation(
	diagnostic: ts.Diagnostic,
): diagnostic is ts.DiagnosticWithLocation {
	return !!(
		diagnostic.file &&
		typeof diagnostic.start === 'number' &&
		typeof diagnostic.length === 'number'
	);
}

/**
 * Get the message for a diagnostic. TypeScript
 * is kind of weird: messageText might have the message,
 * or a pointer to the message. This follows the chain
 * to get a string, regardless of which case we're in.
 */
export function tsDiagnosticMessage(diagnostic: Pick<ts.Diagnostic, 'messageText'>): string {
	if (typeof diagnostic.messageText === 'string') {
		return diagnostic.messageText;
	}
	// TODO: go through linked list
	return diagnostic.messageText.messageText;
}

/**
 * TypeScript and CodeMirror have slightly different
 * ways of representing diagnostics. This converts
 * from one to the other.
 */
export function convertTSDiagnosticToCM(d: ts.DiagnosticWithLocation): Diagnostic {
	// We add some code at the end of the document, but we can't have a
	// diagnostic in an invalid range
	const start = tsPosToCm(d.start);
	const message = tsDiagnosticMessage(d);

	return {
		from: start,
		to: start + d.length,
		message,
		severity: tsCategoryToSeverity(d),
	};
}

function processSchema(schema: Schema): string {
	switch (schema.type) {
		case 'string':
		case 'number':
		case 'boolean':
		case 'bigint':
		case 'symbol':
		case 'null':
		case 'undefined':
			return schema.type;

		case 'function':
			return 'Function';

		case 'array':
			if (Array.isArray(schema.value)) {
				// Handle tuple type if array has different types
				if (schema.value.length > 0) {
					return `Array<${schema.value[0].type}>`;
				}
				return '[]';
			}
			return `${schema.value}[]`;

		case 'object':
			if (!Array.isArray(schema.value)) {
				return '{}';
			}

			const properties = schema.value
				.map((prop) => {
					const key = prop.key ?? 'unknown';
					const type = processSchema(prop);
					return `  ${key}: ${type};`;
				})
				.join('\n');

			return `{\n${properties}\n}`;

		default:
			return 'any';
	}
}

export function schemaToTypescriptTypes(schema: Schema, interfaceName: string): string {
	return `interface ${interfaceName} ${processSchema(schema)}`;
}

function docToTypescriptType(name: string, doc?: DocMetadata) {
	if (!doc) return null;
	return `${name}(${doc?.args?.map((arg) => argToTypescriptType(arg)) ?? ''}): ${doc?.returnType ?? 'any'};`;
}

function argToTypescriptType(arg: DocMetadataArgument) {
	return `${arg.name}${arg.optional ? '?:' : ':'} ${arg.type ?? 'any'}`;
}

export async function generateExtensionTypes() {
	const extensions = await import('n8n-workflow').then((module) => module.ExpressionExtensions);
	const types = extensions
		.map(
			({ typeName, functions }) => `interface ${capitalize(typeName)} {
${Object.entries(functions)
	.map(([name, fn]) => docToTypescriptType(name, fn.doc))
	.filter(Boolean)
	.join('\n')}
}`,
		)
		.join('\n');

	return `export {}
declare global {
	${types}
}`;
}
