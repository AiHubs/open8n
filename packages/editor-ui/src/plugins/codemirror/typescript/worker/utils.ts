import { type CodeExecutionMode } from 'n8n-workflow';

export const fnPrefix = (mode: CodeExecutionMode) => `(
/**
 * @returns {${returnTypeForMode(mode)}}
*/
() => {\n`;

export function wrapInFunction(script: string, mode: CodeExecutionMode): string {
	return `${fnPrefix(mode)}${script}\n})()`;
}

export function globalTypeDefinition(types: string) {
	return `export {};
declare global {
	${types}
}`;
}

export function returnTypeForMode(mode: CodeExecutionMode): string {
	const returnItem = '{ json: { [key: string]: unknown } } | { [key: string]: unknown }';
	if (mode === 'runOnceForAllItems') {
		return `Promise<Array<${returnItem}>> | Array<${returnItem}>`;
	}

	return `Promise<${returnItem}> | ${returnItem}`;
}
