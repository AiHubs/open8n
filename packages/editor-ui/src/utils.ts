import xss, { friendlyAttrValue } from 'xss';
import { Primitives, Optional, JsonSchema, JsonSchemaType } from "@/Interface";

export const omit = (keyToOmit: string, { [keyToOmit]: _, ...remainder }) => remainder;

export function isObjectLiteral(maybeObject: unknown): maybeObject is { [key: string]: string } {
	return typeof maybeObject === 'object' && maybeObject !== null && !Array.isArray(maybeObject);
}

export function isJsonKeyObject(item: unknown): item is {
	json: unknown;
	[otherKeys: string]: unknown;
} {
	if (!isObjectLiteral(item)) return false;

	return Object.keys(item).includes('json');
}

export function sanitizeHtml(dirtyHtml: string) {
	const allowedAttributes = ['href','name', 'target', 'title', 'class', 'id'];
	const allowedTags = ['p', 'strong', 'b', 'code', 'a', 'br', 'i', 'em', 'small' ];

	const sanitizedHtml = xss(dirtyHtml, {
		onTagAttr: (tag, name, value) => {
			if (tag === 'img' && name === 'src') {
				// Only allow http requests to supported image files from the `static` directory
				const isImageFile = value.split('#')[0].match(/\.(jpeg|jpg|gif|png|webp)$/) !== null;
				const isStaticImageFile = isImageFile && value.startsWith('/static/');
				if (!value.startsWith('https://') && !isStaticImageFile) {
					return '';
				}
			}

			// Allow `allowedAttributes` and all `data-*` attributes
			if(allowedAttributes.includes(name) || name.startsWith('data-')) return `${name}="${friendlyAttrValue(value)}"`;

			return;
			// Return nothing, means keep the default handling measure
		},
		onTag: (tag) => {
			if(!allowedTags.includes(tag)) return '';
			return;
		},
	});

	return sanitizedHtml;
}

export const isEmpty = (value?: unknown): boolean => {
	if (!value && value !== 0) return true;
	if(Array.isArray(value)){
		if(!value.length) return true;
		return value.every(isEmpty);
	}
	if (typeof value === 'object') {
		return Object.values(value).every(isEmpty);
	}
	return false;
};

export const intersection = <T>(...arrays: T[][]): T[] => {
	const [a, b, ...rest] = arrays;
	const ab = a.filter(v => b.includes(v));
	return [...new Set(rest.length ? intersection(ab, ...rest) : ab)];
};

export const isObj = (obj: unknown): obj is object => !!obj && Object.getPrototypeOf(obj) === Object.prototype;

export const isSchemaTypeObjectOrList = (type: string) => ['object', 'list'].includes(type);

export const getTypeof = (value: unknown): JsonSchemaType => value === null
	? 'null'
	: value instanceof Date
		? 'date'
		: Array.isArray(value)
			? 'list'
			: typeof value;

export const getJsonSchema = (input: Optional<Primitives | object>, key?: string, path = ''): JsonSchema => {
	let schema:JsonSchema = { type: 'undefined', value: 'undefined', path };
	switch (typeof input) {
		case 'object':
			if (input === null) {
				schema = { type: 'string', value: '[null]', path };
			} else if (input instanceof Date) {
				schema = { type: 'date', value: input.toISOString(), path };
			} else if (Array.isArray(input)) {
				schema = {
					type: 'list',
					value: '',
					path: `${path}[*]`,
				};
				const firstItem = input[0];
				if(Array.isArray(firstItem)){
					schema.value = [getJsonSchema(firstItem, '', `${path}[*]`)];
				} else if(isObj(firstItem)){
					schema.value = getJsonSchema(firstItem, '', `${path}[*]`).value;
				} else {
					schema.value = getTypeof(firstItem);
				}
			} else if (isObj(input)) {
				schema = {
					type: 'object',
					value: Object.entries(input).map(([k, v]) => ({ key: k, ...getJsonSchema(v, k, path + `["${k}"]`)})),
					path,
				};
			}
			break;
		case 'string':
			schema = { type: 'string', value: `"${input}"`, path };
			break;
		case 'function':
			schema =  { type: 'function', value: ``, path };
			break;
		default:
			schema =  { type: typeof input, value: String(input), path };
	}

	if (!isSchemaTypeObjectOrList(schema.type)) {
		schema.path = path;
		if(key){
			schema.value = schema.type;
		}
	}

	return schema;
};
