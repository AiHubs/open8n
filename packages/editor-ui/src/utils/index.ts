export * from './canvasUtils';
export * from './generalUtils';
