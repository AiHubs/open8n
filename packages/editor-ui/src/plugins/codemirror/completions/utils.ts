import { NODE_TYPES_EXCLUDED_FROM_AUTOCOMPLETION } from '@/components/CodeNodeEditor/constants';
import { SPLIT_IN_BATCHES_NODE_TYPE } from '@/constants';
import { useWorkflowsStore } from '@/stores/workflows';
import { resolveParameter } from '@/mixins/workflowHelpers';

export function autocompletableNodeNames() {
	return useWorkflowsStore()
		.allNodes.filter((node) => !NODE_TYPES_EXCLUDED_FROM_AUTOCOMPLETION.includes(node.type))
		.map((node) => node.name);
}

// @TODO: Refactor to take two args
export const longestCommonPrefix = (strings: string[]) => {
	if (strings.length === 0) return '';

	return strings.reduce((acc, next) => {
		let i = 0;

		while (acc[i] && next[i] && acc[i] === next[i]) {
			i++;
		}

		return acc.slice(0, i);
	});
};

/**
 * Move selected elements to the start of an array, in order.
 * Selected elements are assumed to be in the array.
 */
export function bringForward(array: string[], selected: string[]) {
	const copy = [...array];

	[...selected].reverse().forEach((s) => {
		const index = copy.indexOf(s);

		if (index !== -1) copy.unshift(copy.splice(index, 1)[0]);
	});

	return copy;
}

/**
 * Whether a string may be used as a key in object dot notation access.
 */
export const isAllowedInDotNotation = (str: string) => {
	const DOT_NOTATION_BANNED_CHARS = /^(\d)|[\\ `!@#$%^&*()_+\-=[\]{};':"\\|,.<>?~]/g;

	return !DOT_NOTATION_BANNED_CHARS.test(str);
};

export const isSplitInBatchesAbsent = () =>
	!useWorkflowsStore().workflow.nodes.some((node) => node.type === SPLIT_IN_BATCHES_NODE_TYPE);

export const receivesNoBinaryData = () => resolveParameter('={{ $binary }}')?.data === undefined;

export function hasNoParams(toResolve: string) {
	const PSEUDO_PARAMS = ['notice']; // not proper params, no user input allowed

	const params = resolveParameter(`={{ ${toResolve}.params }}`);

	if (!params) return true;

	const paramKeys = Object.keys(params);

	return paramKeys.length === 1 && PSEUDO_PARAMS.includes(paramKeys[0]);
}
