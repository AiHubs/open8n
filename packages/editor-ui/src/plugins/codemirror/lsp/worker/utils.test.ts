import { generateExtensionTypes, schemaToTypescriptTypes } from './utils';

describe('typescript worker utils', () => {
	describe('generateExtensionTypes', () => {
		it('should work', async () => {
			expect(await generateExtensionTypes()).toMatchSnapshot();
		});
	});

	describe('schemaToTypescriptTypes', () => {
		it('should work', () => {
			expect(
				schemaToTypescriptTypes(
					{
						type: 'object',
						value: [
							{
								key: 'test',
								type: 'string',
								value: '',
								path: '.test',
							},
							{
								type: 'object',
								key: 'nested',
								path: '.nested',
								value: [
									{
										key: 'amount',
										type: 'number',
										value: '',
										path: '.amount',
									},
								],
							},
							{
								type: 'array',
								key: 'nestedArray',
								path: '.nestedArray',
								value: [
									{
										type: 'object',
										key: 'nested',
										path: '.nestedArray.nested',
										value: [
											{
												key: 'amount',
												type: 'number',
												value: '',
												path: '.amount',
											},
										],
									},
								],
							},
						],
						path: '',
					},
					'Node name 1',
				),
			).toEqual(`interface NodeName_1 {
  test: string;
  nested: {
  amount: number;
};
  nestedArray: Array<{
  amount: number;
}>;
}`);
		});
	});
});
