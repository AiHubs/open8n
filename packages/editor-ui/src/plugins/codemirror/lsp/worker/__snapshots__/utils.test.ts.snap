// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typescript worker utils > generateExtensionTypes > should work 1`] = `
"export {}
declare global {
	interface Array {
removeDuplicates(fieldNames: any): any;
unique(fieldNames: any): any;
first(): any;
last(): any;
pluck(fieldNames: string): Array;
randomItem(): any;
sum(): number;
min(): number;
max(): number;
average(): number;
isNotEmpty(): boolean;
isEmpty(): boolean;
compact(): Array;
smartJoin(keyField: string,nameField: string): Object;
chunk(length: number): Array;
renameKeys(from: string,to: string): Array;
merge(otherArray: Array): Object;
union(otherArray: Array): Array;
difference(otherArray: Array): Array;
intersection(otherArray: Array): Array;
append(elements: any): Array;
toJsonString(): string;
}
interface Date {
beginningOf(unit?: DurationUnit): DateTime;
endOfMonth(): DateTime;
extract(unit?: string): number;
isBetween(date1: string | DateTime,date2: string | DateTime): boolean;
isDst(): boolean;
isInLast(n: number,unit?: DurationUnit): boolean;
isWeekend(): boolean;
minus(n: number | object,unit?: string): DateTime;
plus(n: number | object,unit?: string): DateTime;
format(fmt: string): string;
toDateTime(): DateTime;
diffTo(otherDateTime: string | DateTime,unit: string | string[]): number | Record<DurationUnit, number>;
diffToNow(unit: string | string[]): number | Record<DurationUnit, number>;
isEmpty(): boolean;
isNotEmpty(): boolean;
}
interface Number {
ceil(): number;
floor(): number;
format(locale?: string,options?: object): string;
round(decimalPlaces?: number): number;
abs(): number;
isInteger(): boolean;
isEven(): boolean;
isOdd(): boolean;
toBoolean(): boolean;
toDateTime(format?: string): DateTime;
}
interface Object {
isEmpty(): boolean;
isNotEmpty(): boolean;
hasField(name: string): boolean;
removeField(key: string): Object;
removeFieldsContaining(value: string): Object;
keepFieldsContaining(value: string): Object;
compact(): Object;
urlEncode(): string;
keys(): Array;
values(): Array;
toJsonString(): string;
}
interface String {
hash(algo?: string): string;
removeMarkdown(): string;
removeTags(): string;
toDate(): Date;
toDateTime(format?: string): DateTime;
toBoolean(): boolean;
toNumber(): number;
toFloat(): number;
toInt(radix?: number): number;
toSentenceCase(): string;
toSnakeCase(): string;
toTitleCase(): string;
urlDecode(allChars?: boolean): string;
urlEncode(allChars?: boolean): string;
quote(mark?: string): string;
replaceSpecialChars(): string;
length(): number;
isDomain(): boolean;
isEmail(): boolean;
isNumeric(): boolean;
isUrl(): boolean;
isEmpty(): boolean;
isNotEmpty(): boolean;
extractEmail(): string;
extractDomain(): string;
extractUrl(): string;
extractUrlPath(): string;
parseJson(): any;
base64Encode(): string;
base64Decode(): string;
}
interface Boolean {
toNumber(): number;
}
}"
`;
